packages/grafana-data/src/types/datasource.ts

```diff
@@ -5,18 +5,32 @@ import { PanelData } from './panel';
 import { LogRowModel } from './logs';
 import { AnnotationEvent, KeyValue, LoadingState, TableData, TimeSeries } from './data';
 import { DataFrame, DataFrameDTO } from './dataFrame';
-import { RawTimeRange, TimeRange } from './time';
+import { RawTimeRange, TimeRange, AbsoluteTimeRange } from './time';
 import { ScopedVars } from './ScopedVars';
 
 export interface DataSourcePluginOptionsEditorProps<JSONData = DataSourceJsonData, SecureJSONData = {}> {
   options: DataSourceSettings<JSONData, SecureJSONData>;
   onOptionsChange: (options: DataSourceSettings<JSONData, SecureJSONData>) => void;
 }
 
+export type DataSourceQuery<DSType extends DataSourceApi<any, any>> = DSType extends DataSourceApi<
```

Comment:
@kaydelaney can you say something more about these changes? Can't see how they fit into TS upgrade.

cc @ryantxu

Reply:
Sure. Typescript 3.7 is somewhat more strict with its typechecking in certain scenarios, which caused a type error when doing the following
```typescript
const plugin = new DataSourcePlugin(lokiDatasource);
```
with the error
```
"Argument of type 'typeof LokiDatasource' is not assignable to parameter of type 'DataSourceConstructor<DataSourceApi<DataQuery, LokiOptions>, DataQuery, LokiOptions>'."
```
So in order for the correct types to be properly inferred the helper types above were introduced to extract the `TQuery` and `TOptions` types of a type extending `DataSourceApi<TQuery, TOptions>`.